File: atcoder/ahc/ahc017/a.py
import time
import random
import sys
from collections import deque
input = sys.stdin.readline

stime = time.time()
TIME_LIM = 6
inf = 10**9

center = 1000
SPLIT = 10

n,m,d,k = map(int,input().split())

UVW = [list(map(int,input().split())) for i in range(m)]
UVW = [[u-1,v-1,w] for u,v,w in UVW]
XY = [list(map(int,input().split())) for i in range(n)]

g = [[] for i in range(n)]
for i,(u,v,w) in enumerate(UVW):
    g[u].append((v,w,i))
    g[v].append((u,w,i))

xma = [-inf,-1]
xmi = [inf,-1]
yma = [-inf,-1]
ymi = [inf,-1]
# cen = [inf,-1]

for i,(x,y) in enumerate(XY):
    if xma[0] < x:
        xma = [x,i]
    if xmi[0] > x:
        xmi = [x,i]
    
    if yma[0] < y:
        yma = [y,i]
    if ymi[0] > y:
        ymi = [y,i]
    
    # r = (x-500)**2+(y-500)**2
    # if cen[0] > r:
    #     cen = [r,i]

dist_point = set([xma[1],xmi[1],yma[1],ymi[1]])

# dist_point = set([xma[1],xmi[1],yma[1],ymi[1],cen[1]])
# dist_lis = [inf]*m
def dist_sum(used,day=1):
    
    dist_count = 0
    for i in dist_point:
        dist = [inf]*n
        dist[i] = 0
        q = deque([i])
        while q:
            now = q.popleft()
            for nex,w,id in g[now]:
                if used[id]:
                    continue
                if dist[nex] > dist[now]+w:
                    dist[nex] = dist[now]+w
                    q.append(nex)
        dist_count += sum(dist)

    return dist_count

class Unionfind:
     
    def __init__(self,n):
        self.uf = [-1]*n
 
    def find(self,x):
        if self.uf[x] < 0:
            return x
        else:
            self.uf[x] = self.find(self.uf[x])
            return self.uf[x]
 
    def same(self,x,y):
        return self.find(x) == self.find(y)
 
    def union(self,x,y):
        x = self.find(x)
        y = self.find(y)
        if x == y:
            return False
        if self.uf[x] > self.uf[y]:
            x,y = y,x
        self.uf[x] += self.uf[y]
        self.uf[y] = x
        return True
 
    def size(self,x):
        x = self.find(x)
        return -self.uf[x]


def make_base_R():
    base_R = []
    for i in range(1,d+1):
        need = m//d
        if m%d >= i:
            need += 1
        for j in range(need):
            base_R.append(i)
    return base_R

def pos_split_R():


    nXY = [[x*2-center,y*2-center] for x,y in XY]
    radius = [[] for i in range(SPLIT)]
    lim = [(100*i)**2 for i in range(1,SPLIT+1)]
    for i,(u,v,_) in enumerate(UVW):
        x,y = nXY[u]
        nx,ny = nXY[v]
        cx,cy = (x+nx)//2,(y+ny)//2

        rxy = cx**2+cy**2
        if i < 10:
            print(u,v,cx,cy,rxy)
        for j,li in enumerate(lim):
            if rxy <= li:
                radius[j].append(i)
                break
    
    pos_R = [-1]*m

    now = 0

    for i in range(SPLIT):
        rad = radius[i]
        random.shuffle(rad)

        for x in rad:
            pos_R[x] = now+1
            now = (now+1)%d
    return pos_R

def pos_split_R_clime():
    
    nXY = [[x*2-center,y*2-center] for x,y in XY]
    radius = [[] for i in range(SPLIT)]
    lim = [(100*i)**2 for i in range(1,SPLIT+1)]
    for i,(u,v,_) in enumerate(UVW):
        x,y = nXY[u]
        nx,ny = nXY[v]
        cx,cy = (x+nx)//2,(y+ny)//2

        rxy = cx**2+cy**2
        if i < 10:
            print(u,v,cx,cy,rxy)
        for j,li in enumerate(lim):
            if rxy <= li:
                radius[j].append((cx,cy,i))
                break
    
    pos_R = [-1]*m

    now = 0

    pos_X = [0]*d
    pos_Y = [0]*d
    num_d = [0]*d
    for i in range(SPLIT):
        rad = radius[i]
        print(len(rad))
        random.shuffle(rad)

        for cx,cy,ind in rad:
            pos_R[ind] = now+1
            pos_X[now] += cx
            pos_Y[now] += cy
            num_d[now] += 1
            now = (now+1)%d

        # print(pos_X,pos_Y,num_d)
        upd = 0
        si = len(rad)
        while upd < si:
            a,b = random.randint(0,si-1),random.randint(0,si-1)
            ax,ay,aind = rad[a]
            bx,by,bind = rad[b]
            d1,d2 =  pos_R[aind]-1,pos_R[bind]-1
            if d1 == d2:
                continue
            
            dx,dy = ax-bx,ay-by
            base_size = (pos_X[d1]/num_d[d1])**2+(pos_Y[d1]/num_d[d1])**2+(pos_X[d2]/num_d[d2])**2+(pos_Y[d2]/num_d[d2])**2
            
            change_size = ((pos_X[d1]-dx)/num_d[d1])**2+((pos_Y[d1]-dy)/num_d[d1])**2+((pos_X[d2]+dx)/num_d[d2])**2+((pos_Y[d2]+dy)/num_d[d2])**2

            if base_size <= change_size:
                upd += 1
                continue

            upd = 0
            pos_X[d1] -= dx
            pos_Y[d1] -= dy
            pos_X[d2] += dx
            pos_Y[d2] += dy
            pos_R[aind],pos_R[bind] = pos_R[bind],pos_R[aind]

        # print(pos_X,pos_Y,num_d)
    return pos_R


def out(R):
    print(*R)
    exit()

class DayInfo:

    def __init__(self,road):
        self.road = road
        self.uf = Unionfind(n)
        self.tree_created = False
        self.spanning_tree = [0]*m
        self.inf_path_num = 0
        self.upd = False
        self.dist_sum = inf
        self._build()



    def _build(self):
        road_info = [[x,i] for i,x in enumerate(self.road)]
        random.shuffle(road_info)
        for use,ind in road_info:
            if use:
                continue
            u,v,_ = UVW[ind]
            if self.uf.same(u,v):
                continue
            self.uf.union(u,v)
            self.spanning_tree[ind] = 1
        
        if self.uf.size(0) == n:
            self.tree_created = True

        else:
            for i in range(n):
                if self.uf.find(i) == i:
                    si = self.uf.size(i)
                    self.inf_path_num += si*(n-si)
    
    def rebuild(self,ind=None):

        self.spanning_tree = [0]*m
        self.uf = Unionfind(n)
        self.tree_created = False
        road_info = [[x,i] for i,x in enumerate(self.road)]
        random.shuffle(road_info)
        if ind != None:
            for i in range(m):
                if road_info[i][1] == ind:
                    road_info = road_info[:i] + road_info[i+1:] + [road_info[i]]
                    break

        for use,ind in road_info:
            if use:
                continue
            u,v,_ = UVW[ind]
            if self.uf.same(u,v):
                continue
            self.uf.union(u,v)
            self.spanning_tree[ind] = 1
        
        if self.uf.size(0) == n:
            self.tree_created = True

        else:
            self.inf_path_num = 0
            for i in range(n):
                if self.uf.find(i) == i:
                    si = self.uf.size(i)
                    self.inf_path_num += si*(n-si)

    def calc_dist(self):
        if self.upd:
            return self.dist_sum
        self.dist_sum = dist_sum(self.road)
        self.upd = True
        return self.dist_sum

    def dist_set(self,dist):
        assert self.upd
        self.dist_sum = dist


class Master:

    def __init__(self):
        # self.base_R = make_base_R()
        # random.shuffle(self.base_R)
        self.base_R = pos_split_R_clime()
        self.DayInfos = []
        self.has_load = [set() for i in range(d)]
        self._build()
        self.dist_score = 10**20
        self.upd_score = False

    def _build(self):
        days_R = [[0]*m for i in range(d)]
        for i,r in enumerate(self.base_R):
            days_R[r-1][i] = 1
            self.has_load[r-1].add(i)

        for day_R in days_R:
            self.DayInfos.append(DayInfo(day_R))

    
    def check_tree(self):
        for dayInfo in self.DayInfos:
            if dayInfo.tree_created == False:
                return False
        return True

    def check_inf_num(self):
        inf_num = 0
        for dayInfo in self.DayInfos:
            inf_num += dayInfo.inf_path_num
        return inf_num

    def make_tree(self):
        if self.check_tree():
            return True

        for i,r in enumerate(self.base_R):
            u,v,_ = UVW[i]

            ### if already same group, continue
            if self.DayInfos[r-1].uf.same(u,v):
                continue

            ### if group is different, try to swap with other dayinfo

            s = random.randint(0,d-1)
            for j in range(d):
                ss = (j+s)%d
                if ss == r-1:
                    continue
                if self.DayInfos[ss].spanning_tree[i]:
                    continue
                
                find_edge = -1
                for edge in self.has_load[ss]:
                    if self.DayInfos[r-1].spanning_tree[edge]:
                        continue
                    find_edge = edge
                    break

                if find_edge == -1:
                    continue

                self.base_R[edge],self.base_R[i] = r,ss+1
                
                self.DayInfos[r-1].spanning_tree[i] = 1
                self.has_load[r-1].remove(i)
                self.has_load[r-1].add(edge)
                self.has_load[ss].remove(edge)
                self.has_load[ss].add(i)
                self.DayInfos[r-1].road[edge],self.DayInfos[r-1].road[i] = 1,0
                self.DayInfos[ss].road[edge],self.DayInfos[ss].road[i] = 0,1
                break
        self.upd_score = False

        if self.check_tree():
            return True
        else:
            return False

    def make_tree_hard(self):
        if self.check_tree():
            return True

        for i,r in enumerate(self.base_R):
            u,v,_ = UVW[i]

            ### if already same group, continue
            if self.DayInfos[r-1].uf.same(u,v):
                continue

            ### if group is different, try to swap with other dayinfo

            s = random.randint(0,d-1)
            for j in range(d):
                ss = (j+s)%d
                if ss == r-1:
                    continue
                if self.DayInfos[ss].spanning_tree[i]:
                    self.DayInfos[ss].rebuild(i)
                
                if self.DayInfos[ss].spanning_tree[i]:
                    continue

                find_edge = -1
                for edge in self.has_load[ss]:
                    if self.DayInfos[r-1].spanning_tree[edge]:
                        continue
                    find_edge = edge
                    break

                if find_edge == -1:
                    continue

                self.base_R[edge],self.base_R[i] = r,ss+1
                
                self.DayInfos[r-1].spanning_tree[i] = 1
                self.has_load[r-1].remove(i)
                self.has_load[r-1].add(edge)
                self.has_load[ss].remove(edge)
                self.has_load[ss].add(i)
                self.DayInfos[r-1].road[edge],self.DayInfos[r-1].road[i] = 1,0
                self.DayInfos[ss].road[edge],self.DayInfos[ss].road[i] = 0,1
                break
        self.upd_score = False

        if self.check_tree():
            return True
        else:
            return False


    def calc_score(self):
        if self.upd_score:
            return self.dist_score

        self.dist_score = 0
        for dayInfo in self.DayInfos:
            self.dist_score += dayInfo.calc_dist()
        
        self.upd_score = True
        return self.dist_score


    def out(self):
        print(*self.base_R)
        exit()


best_master = Master()
# best_score = best_master.calc_score()
# tree_made = best_master.check_tree()
# inf_num = best_master.check_inf_num()

# TREE_LIM = 0.8
# while time.time() - stime < TREE_LIM:
#     master = Master()
#     made = master.make_tree()
#     # if made == False:
#     #     made = master.make_tree_hard()

#     now_score = master.calc_score()
#     now_inf_num = master.check_inf_num()
#     if tree_made == False and master.check_tree():
#         best_score = now_score
#         best_master = master
#         tree_made = master.check_tree()
#         inf_num = now_inf_num
#         continue

#     if now_inf_num < inf_num:
#         best_score = now_score
#         best_master = master
#         tree_made = master.check_tree()
#         inf_num = now_inf_num
#     elif now_inf_num == inf_num:
#         if now_score < best_score:
#             if tree_made and master.check_tree() == False:
#                 continue
#             best_score = now_score
#             best_master = master
#             tree_made = master.check_tree()

# print(*best_master.base_R)
num1 = 0
num2 = 0
while time.time() - stime < TIME_LIM-0.5:
    num1 += 1
    e1,e2 = random.randint(0,m-1),random.randint(0,m-1)
    d1,d2 = best_master.base_R[e1]-1,best_master.base_R[e2]-1
    DayInfo1 = best_master.DayInfos[d1]
    DayInfo2 = best_master.DayInfos[d2]
    if d1 == d2:
        continue
    # if d1 == d2 or DayInfo1.spanning_tree[e2] or DayInfo2.spanning_tree[e1]:
    #     continue
    num2 += 1
    dist1 = DayInfo1.calc_dist()
    dist2 = DayInfo2.calc_dist()
    base_score = dist1 + dist2
    DayInfo1.road[e1],DayInfo1.road[e2] = 0,1
    DayInfo2.road[e1],DayInfo2.road[e2] = 1,0

    DayInfo1.upd = False
    DayInfo2.upd = False
    new_score = DayInfo1.calc_dist()+DayInfo2.calc_dist()

    if new_score > base_score:
        DayInfo1.road[e1],DayInfo1.road[e2] = 1,0
        DayInfo2.road[e1],DayInfo2.road[e2] = 0,1
        DayInfo1.dist_set(dist1)
        DayInfo2.dist_set(dist2)
    else:
        best_master.base_R[e1],best_master.base_R[e2] = d2+1,d1+1
print(num1,num2)
best_master.out()

File: atcoder/ahc/ahc017/a.cpp
#include <bits/stdc++.h>
#include <chrono>
#include <iostream>
#include <vector>
#include <algorithm>
#include <cassert>
#include <climits>
#include <cmath>

using namespace std;
using namespace chrono;
using ll = long long;
uint64_t rand_seed = 123456789;
uint64_t xorshift64() {
	rand_seed ^= rand_seed << 13;
	rand_seed ^= rand_seed >> 7;
	rand_seed ^= rand_seed << 17;
	return rand_seed;
}
int rand_int(int l, int r) {
	return l + int(xorshift64() % (r - l));
}

float TIME_LIM = 6;
const int center=1000,SPLIT=10,inf=1e9;
int n,m,d,k;
const int MAX = 3005;
int U[MAX],V[MAX], W[MAX],X[MAX],Y[MAX];

struct edge {
  int to, id;
  ll cost;
  edge() {}
  edge(int to, ll cost,int id): to(to), cost(cost),id(id) {}
};

vector<vector<edge> > G(MAX);
ll dist_sum(vector<int> used,vector<int> dist_point,int n)
{
    ll dist_count = 0;
    for (int i=0;i<2;i++){
        vector<ll> dist(n,inf);
        int s = dist_point[i];
        queue<int> que;
        que.push(s);  // sから探索する
        dist[s] = 0;
        while (que.size() != 0) {     // 幅優先探索
            int state = que.front(); que.pop();
            for (auto next : G[state]) {
                if (used[next.id]) continue;
                ll nd = dist[state] + next.cost;
                if (nd < dist[next.to]) {  // 未探索の時のみ行う
                    dist[next.to] = nd;
                    que.push(next.to);  //次の状態をqueueへ格納
                }
            }
        }
        for (int j = 0;j<n;j++){
            dist_count += dist[j];
        }

        
    }
    return dist_count;
}

int main() {
    auto startClock = system_clock::now();
    cin >> n >> m >> d >> k;
    for (int i = 0;i < m;i++){
        cin >> U[i] >> V[i] >> W[i];
        U[i]--;
        V[i]--;
    }  
    pair<int,int> xma = make_pair(-inf,-1);
    pair<int,int> xmi = make_pair(inf,-1);
    pair<int,int> yma = make_pair(-inf,-1);
    pair<int,int> ymi = make_pair(inf,-1);

    for (int i = 0;i<n;i++){
        cin >> X[i] >> Y[i];
        if (xma.first < X[i]){
            xma.first = X[i];
            xma.second = i;
        }
        if (xmi.first >X[i]){
            xmi.first = X[i];
            xmi.second = i;
        }

        if (yma.first < Y[i]){
            yma.first = Y[i];
            yma.second = i;
        }
        if (ymi.first >Y[i]){
            ymi.first = Y[i];
            ymi.second = i;
        }
    }

    vector<int> dist_point = {xma.second,xmi.second,yma.second,ymi.second};
    // vector<int> dist_point = {xma.second,xmi.second};

    for (int i = 0; i < m; i++) {
        int u=U[i],v=V[i],w=W[i];
        G[u].emplace_back(v,w,i);
        G[v].emplace_back(u,w,i);

    }

    // vector<int> nX(n),nY(n);
    // for (int i=0;i<n;i++){
    //     nX[i] = X[i]*2-center;
    //     nY[i] = Y[i]*2-center;
    // }
    // vector<vector<int> > radius(SPLIT);
    // vector<int> lim;
    // for (int i=1;i<=SPLIT;i++){
    //     int l = i*100*i*100;
    //     lim.push_back(l);
    // }
    // for (int i = 0;i<m;i++){
    //     int u = U[i],v = V[i];
    //     int x = nX[u],y=nY[u];
    //     int nx=nX[v],ny=nY[v];

    //     int cx = (x+nx)/2;
    //     int cy = (y+ny)/2;

    //     int rxy = cx*cx+cy*cy;
    //     for (int j = 0;j<SPLIT;j++){
    //         if (rxy <= lim[j]){
    //             radius[j].push_back(i);
    //             break;
    //         }
    //     }
    // }
    // vector<int> pos_R(m);

    // int now = 0;
    // for (int i=0;i<SPLIT;i++){
    //     vector rad = radius[i];
    //     random_device seed_gen;
    //     mt19937 engine(seed_gen());
    //     shuffle(rad.begin(), rad.end(), engine);

    //     for (auto x: rad){
    //         pos_R[x] = now+1;
    //         now = (now+1)%d;
    //     }
    // }

    vector<int> pos_R;

    for (int i = 1; i<=d;i++){
        int need = m/d;
        if (m%d >= i) need++;
        for (int j = 0;j<need;j++){
            pos_R.push_back(i);
        }
    }
    random_device seed_gen;
    mt19937 engine(seed_gen());
    shuffle(pos_R.begin(), pos_R.end(), engine);


    vector<vector<int> > DayInfos(d,vector<int>(m,0));
    for (int i=0;i<m;i++){
        int r = pos_R[i];
        DayInfos[r-1][i] = 1;
    }
  
    // for (int i = 0;i<d;i++){
    //     int num = 0;
    //     for (int j = 0;j<m;j++){
    //         num += DayInfos[i][j];
    //     }
    //     cout << "day" << i+1 << " dist = " << dist_sum(DayInfos[i],dist_point,n) << " " << dist_point.size() << " " << num << endl;
    // }
    vector<int> upd_score(d,0);
    vector<ll> score(d,0);
    const static double END_TIME = TIME_LIM - 0.3;
    int n1=0,n2=0;
    while (true)
    {        
        if (n1%100 == 0){
            const double time = duration_cast<microseconds>(system_clock::now() - startClock).count() * 1e-6;   // 経過時間（秒）
            if (time >= END_TIME)
            {
                break;
            }
        }
        int e1 = rand_int(0,m-1), e2 = rand_int(0,m-1);
        int d1 = pos_R[e1]-1, d2 = pos_R[e2]-1;
        n1++;
        if (d1 == d2) continue;
        n2++;
        ll dist1,dist2;
        if (upd_score[d1]){
            dist1 = score[d1];
        } else{
            dist1 = dist_sum(DayInfos[d1],dist_point,n);
        }
        if (upd_score[d2]){
            dist2 = score[d2];
        } else{
            dist2 = dist_sum(DayInfos[d2],dist_point,n);
        }
 
        ll base_score = dist1+dist2;

        DayInfos[d1][e1] = 0,DayInfos[d1][e2] = 1;
        DayInfos[d2][e1] = 1,DayInfos[d2][e2] = 0;

        ll ndist1 = dist_sum(DayInfos[d1],dist_point,n);
        ll ndist2 = dist_sum(DayInfos[d2],dist_point,n);

        if (ndist1+ndist2 > base_score){
            DayInfos[d1][e1] = 1,DayInfos[d1][e2] = 0;
            DayInfos[d2][e1] = 0,DayInfos[d2][e2] = 1;
            upd_score[d1] = 1;
            upd_score[d2] = 1;
            score[d1] = dist1;
            score[d2] = dist2;
        } else{
            upd_score[d1] = 1;
            upd_score[d2] = 1;
            score[d1] = ndist1;
            score[d2] = ndist2;
            pos_R[e1] = d2+1;
            pos_R[e2] = d1+1;

        }
    }
    // cout << n1 << " " << n2 << endl;
    // for (int i = 0;i<d;i++){
    //     int num = 0;
    //     for (int j = 0;j<m;j++){
    //         num += DayInfos[i][j];
    //     }
    //     cout << "day" << i+1 << " dist = " << dist_sum(DayInfos[i],dist_point,n) << " " << dist_point.size() << " " << num << endl;
    // }

    for (int i=0;i<m;i++){
        cout << pos_R[i] << " ";
    }
    cout << endl;

    return 0;
}